File: src/main/java/chapter2obervables/Tutorial2_2ObserverInterface.java
package chapter2obervables;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

public class Tutorial2_2ObserverInterface {

    public static void main(String[] args) {

        testObserver();
        testObserverLambda();
    }


    private static void testObserver() {

        Observable<String> source =
                Observable.just("Alpha", "Beta", "Gamma", "Delta",
                        "Epsilon");

        Observer<Integer> myObserver = new Observer<Integer>() {
            @Override
            public void onSubscribe(Disposable d) {
                //do nothing with Disposable, disregard for now
            }

            @Override
            public void onNext(Integer value) {
                System.out.println("RECEIVED: " + value);
            }

            @Override
            public void onError(Throwable e) {
                e.printStackTrace();
            }

            @Override
            public void onComplete() {
                System.out.println("Done!");
            }
        };

        source
                .map(String::length)
                .filter(i -> i >= 5)
                .subscribe(myObserver);
    }

    /**
     * This method accepts 3 lambdas the <strong>onNext</strong> lambda, the <strong>onError</strong> lambda
     * , and the <strong>onComplete</strong> lambda.
     */
    private static void testObserverLambda() {
        Observable<String> source = Observable.just("Alpha", "Beta", "Gamma", "Delta", "Epsilon");

        // INFO Lambdas for onNext, onError, onComplete methods
//        Consumer<Integer> onNext = i -> System.out.println("RECEIVED: " + i);
//        Consumer<Throwable> onError = Throwable::printStackTrace;
//        Action onComplete = () -> System.out.println("Done!");

        source.map(String::length).filter(i -> i >= 5)
                .subscribe(
                        // onNext
                        i -> System.out.println("RECEIVED: " + i),
                        // onError
                        Throwable::printStackTrace,
                        // onComplete
                        () -> System.out.println("Done!")
                );
    }

}

File: src/main/kotlin/chapter2observables/Tutorial2_2ObserverInterface.kt
package chapter2observables

import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.disposables.Disposable

fun main() {


    testObserverWithMultipleObservables()
}

private fun testObserver() {

    val source = Observable.just(
        "Alpha", "Beta", "Gamma", "Delta", "Epsilon"
    )

    val myObserver = object : Observer<String> {


        override fun onSubscribe(d: Disposable) {
            //do nothing with Disposable, disregard for now
        }

        override fun onNext(value: String) {
            println("RECEIVED: $value")
        }

        override fun onError(e: Throwable) {
            e.printStackTrace()
        }

        override fun onComplete() {
            println("Done!")
        }
    }

    source.subscribe(myObserver)


}


/**
 * This method accepts 3 lambdas the **onNext** lambda, the **onError** lambda
 * , and the **onComplete** lambda.
 */
private fun testObserverLambda() {
    val source = Observable.just("Alpha", "Beta", "Gamma", "Delta", "Epsilon")

    source
        .map {
            it.length
        }
        .filter { it >= 5 }
        .subscribe(
            {
                println("RECEIVED $it")
            },
            {
                it.printStackTrace()
            },
            {
                println("onComplete()")
            }
        )

}

private fun testObserverWithMultipleObservables() {
    val source1 = Observable.just(
        "Alpha", "Beta", "Gamma", "Delta", "Epsilon"
    )

    val source2 = Observable.just("One", "Two", "Three", "Four")

    val myObserver = object : Observer<String> {


        override fun onSubscribe(d: Disposable) {
            //do nothing with Disposable, disregard for now
        }

        override fun onNext(value: String) {
            println("RECEIVED: $value")
        }

        override fun onError(e: Throwable) {
            e.printStackTrace()
        }

        override fun onComplete() {
            println("Done!")
        }
    }

    source1.subscribe(myObserver)
    source2.subscribe(myObserver)

    /*
        Prints:
        RECEIVED: Alpha
        RECEIVED: Beta
        RECEIVED: Gamma
        RECEIVED: Delta
        RECEIVED: Epsilon
        Done!
        RECEIVED: One
        RECEIVED: Two
        RECEIVED: Three
        RECEIVED: Four
        Done!


     */
}